src/herring/model.hpp

```diff
@@ -0,0 +1,275 @@
+    using lookup_tree_t = lookup_tree<value_t, feature_index_t, offset_t, output_index_t, output_t>;
+    using tree_type = std::conditional_t<
+      std::is_same_v<value_t, output_t> || std::is_same_v<output_index_t, output_t>,
+      simple_tree_t,
+      lookup_tree_t
+    >;
+    using sum_elem_type = typename is_container_specialization<output_t, std::vector>::value_type;
+
+    std::vector<tree_type> trees;
+    std::size_t num_class;
+    std::size_t num_feature;
+    row_op row_postproc;
+    float average_factor;
+    float bias;
+    float postproc_constant;
+    std::vector<bool> mutable row_has_missing;
+    bool use_inclusive_threshold;
+
+    void predict(float const* input, std::size_t num_row, float* output, thread_count<int> nthread) const {
+      if (!precompute_missing(input, num_row)) {
+        if (!use_inclusive_threshold) {
+          predict_<false, false>(input, output, num_row, nthread);
+        } else {
+          predict_<false, true>(input, output, num_row, nthread);
+        }
+      } else {
+        if (!use_inclusive_threshold) {
+          predict_<true, false>(input, output, num_row, nthread);
+        } else {
+          predict_<true, true>(input, output, num_row, nthread);
+        }
+      }
+    }
+
+    void set_element_postproc(element_op element_postproc) {
+      postprocess_element = [this, element_postproc]() -> std::function<void(sum_elem_type, float*)> {
+        auto constant = postproc_constant;
+        switch(element_postproc) {
+          case element_op::signed_square:
+            return [](sum_elem_type elem, float* out) {
+              *out = std::copysign(elem * elem, elem);
+            };
+          case element_op::hinge:
+            return [](sum_elem_type elem, float* out) {
+              *out = elem > sum_elem_type{} ? sum_elem_type{1} : sum_elem_type{0};
+            };
+          case element_op::sigmoid:
+            return [constant](sum_elem_type elem, float* out) {
+              *out = sum_elem_type{1} / (sum_elem_type{1} + std::exp(-constant * elem));
+            };
+          case element_op::exponential:
+            return [](sum_elem_type elem, float* out) {
+              *out = std::exp(elem);
+            };
+          case element_op::exponential_standard_ratio:
+            return [constant](sum_elem_type elem, float* out) {
+              *out = std::exp(-elem / constant);
+            };
+          case element_op::logarithm_one_plus_exp:
+            return [](sum_elem_type elem, float* out) {
+              *out = std::log1p(std::exp(elem));
+            };
+          default:
+            return [](sum_elem_type elem, float* out) {
+              *out = elem;
+            };
+        }
+      }();
+    }
+
+   private:
+    std::function<void(sum_elem_type, float*)> postprocess_element;
+
+    auto precompute_missing(float const* input, std::size_t num_row) const {
```

Comment:
Good thought! I suspect a `std::execution::par` here would give us a bit of a boost for low effort. I'm going to go ahead and make an issue out of it, though, because I'm fearful of introducing a perf regression without running all the benchmarks again.

Reply:
https://github.com/triton-inference-server/fil_backend/issues/211